### DISCUSSION TITLE
AttributeError: 'Activation' object has no attribute 'get_output_at'

### DISCUSSION THREAD

**Discussion by NishaAkole** [mindee/doctr#1511] [Q&A]
I am trying to train the custom model on my dataset. Earlier I used PyTorch version of detection model and there were some errors in the script I could not resolve. So, I tried Tensorflow version, and stuck at this error on line no 134 in tensorflow.py script.
![Screenshot (20)](https://github.com/mindee/doctr/assets/47945301/e9a1ed1b-649a-4e56-851b-327c66925403)

**Comment by felixdittrich92** [comment]
Hi @NishaAkole 👋 ,

docTR requires TF <2.16 atm but we have the upgrade on track in #1461

**Reply by felixdittrich92** [reply]
Which errors have you had with pytorch ?

**Reply by NishaAkole** [reply]
I am stuck at Detection model only. The screenshot of the error is attached here. I have NVIDIA RTX 4070 8GB but could not access my GPU by both PyTorch and Tensorflow . Because of that I was facing memory errors as well, so I reduced the dataset size, training set (8 images) and val set(5 images). Then these errors started coming. 
![Screenshot (22)](https://github.com/mindee/doctr/assets/47945301/48fa1aa1-8786-4f52-959e-9cfebe453163)

**Reply by felixdittrich92** [reply]
@NishaAkole This looks like your annotations are wrong

You can use the following as reference :) (polygon coord order doesn't matter)
FUNSD dataset (with doctr annotations)
https://drive.google.com/file/d/1CHVd2B4J4ejbwum0SCV5xD8Flj2UOqNX/view?usp=drive_link

**Reply by NishaAkole** [reply]
Hi @felixdittrich92,
Thanks for your suggestions. 
I tried downgrading Tensoflow to 2.15.1, but still getting the same error. 

I also changed the annotation format and made it exactly same as you mentioned, now getting MemoryError. I will attach the screenshot here. I have GPU NVIDIA RTX 4070, but neither PyTorch nor TensorFlow was able to access it.  
![Screenshot (23)](https://github.com/mindee/doctr/assets/47945301/e52aa88f-b1d4-4c70-9a7e-f7e270462a0f)

**Comment by BirenMer** [comment]
Had same error and it was fixed by downgrading tensorflow version from 2.16.1 to 2.15.1
`pip install tensorflow==2.15.1`

**Reply by LakshX413** [reply]
Thanks! This worked for me.

**Comment by ErvinSabic** [comment]
I just did a whole lot of driver updates so this is currently not viable for me without back-tracking my drivers, CUDA, & TensorRT install (Which is kind of a pain, I just did a kernel update too while I was doing all of that to prep my PC for this avenue.)

What is the status on this update? I'm sitting on TF 2.18 currently. The older compatible version won't install for me, I end up with the following errors: 
```
INFO: pip is looking at multiple versions of tensorflow[and-cuda] to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement tensorrt-libs==8.6.1; extra == "and-cuda" (from tensorflow[and-cuda]) (from versions: 9.0.0.post11.dev1, 9.0.0.post12.dev1, 9.0.1.post11.dev4, 9.0.1.post12.dev4, 9.1.0.post11.dev4, 9.1.0.post12.dev4, 9.2.0.post11.dev5, 9.2.0.post12.dev5, 9.3.0.post11.dev1, 9.3.0.post12.dev1)
ERROR: No matching distribution found for tensorrt-libs==8.6.1; extra == "and-cuda"
```

I am currently running Python 3.11. 

Is there anything we can do to help move it forward? Maybe a work-around in the meantime? 

Is this PR related to the update? - https://github.com/mindee/doctr/pull/1724

Thank you for everything you do.

**Reply by felixdittrich92** [reply]
Hi @NishaAkole :wave:,

In general if possible I would suggest to use PyTorch as docTR backend (which will also be the default backend with the next release).

Otherwise which OS do you use ?

A workaround would be to do:

```
pip3 install tensorflow tf-keras tf2onnx
pip3 install python-doctr  # without any specificer like python-doctr[tf]
# Or if locally cloned
cd doctr & pip3 install -e .
```

Currently docTR still requires Keras v2 instead of Keras v3 the default since tensorflow 2.16, because there are lots of issues which breaks all our models ..so we did some preparation already to migrate everything to keras v3 but it's not fully done.

Additional to be save to use Keras v2 you can put the following on top of your script before any other imports:

```
from doctr.file_utils import ensure_keras_v2

ensure_keras_v2()
```

Best,
Felix :)

**Reply by ErvinSabic** [reply]
This is an excellent work-around that I will use in the meantime. I didn't think the current TF version was compatible with Keras V2. 

Thank you Felix.
